reconciler/healthchecks/healthchecks.go

```diff
@@ -259,57 +253,49 @@ func (s *healthCheckState) performHealthCheck() {
 		Timeout:   timeout,
 	}
 
-	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", s.serverIP, checkURL.Port(), checkURL.Path))
+	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", c.serverIP, checkURL.Port(), checkURL.Path))
 	if err != nil {
 		panic(err)
 	}
 
 	resp, err := client.Get(serverURL.String())
 	if err != nil {
 		log.Infof("%s inaccessible: %v", serverURL, err)
-		s.incrementDown()
+		c.incrementDown()
 		return
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode < 200 || 300 <= resp.StatusCode {
 		body, _ := ioutil.ReadAll(resp.Body)
 		log.Infof("%s returned %d: %s", serverURL, resp.StatusCode, string(body))
-		s.incrementDown()
+		c.incrementDown()
 		return
 	}
-	s.incrementUp()
+	c.incrementUp()
 }
 
-func (s *healthCheckState) incrementUp() {
-	s.Lock()
-	defer s.Unlock()
-	status := s.status
-	switch status.state {
-	case ServerDown:
-		status.count++
-		if status.count >= s.healthCheck.UpThreshold {
-			status.state = ServerUp
-			status.count = 0
-			s.transitionFn(status.state)
-		}
-	case ServerUp:
-		status.count = 0
-	}
+func (c *check) incrementUp() {
+	c.state.increment(ServerDown, ServerUp, c.healthCheck.UpThreshold)
 }
 
-func (s *healthCheckState) incrementDown() {
+func (c *check) incrementDown() {
+	c.state.increment(ServerUp, ServerDown, c.healthCheck.DownThreshold)
+}
+
+func (s *checkState) increment(current, next ServerStatus, transitionThreshold uint32) {
```

Comment:
I think this method would be clearer like this:
```
increment(target,  current, threshold {
switch  s.status {
case current:
    transitionStuff
case target:
   reset count
}
```
It would now read something like:
- increment(Down, Up, 2), _Increment (the target state) Down (and transition if it is currently) Up (with a threshold of) 2_
increment(Up, Down, 2),  _Increment (the target state) Up (and transition if it is currently) Down (with a threshold of) 2_
```